### ISSUE TITLE
Secrets are not namespaced with base.namespace when generated with secretGenerator

### ISSUE THREAD

**Issue by yanc0** [kubernetes-sigs/kustomize#87]
In order to reproduce this:

* Create a `kustomization.yaml` file in base directory with `namespace: foo`

```
namespace: foo-bar
resources:
- foo.yaml
```

* Create a `kustomization.yaml` file in overlay directory with secretGenerator

```
bases:
- "../../base"
secretGenerator:
- name: my-secret
  commands:
    config.secret: "uname -a"
  type: "Opaque"
```

**Result:** generated secret is not namespaced in `foo-bar`

**Workaround**: set also namespace in overlayed `kustomization.yaml`

**Comment by sethpollack** [comment]
I think that is intentional - the overlay overrides the base, so it wouldn't make sense for the settings in the base to apply to the overlay.

**Comment by sethpollack** [comment]
Although I can see an argument for `namespace`, `namePrefix`,  `commonLabels`, and `commonAnnotations` to propagate up if not set in the overlay.

**Comment by yanc0** [comment]
Hi @sethpollack,

I understand that overlay take the precedence over base file. But if overlay does not reference any namespace field, the namespace is taken from the base (if set). No?

All patches loaded in my various overlays are getting the right namespace, only referenced in base/kustomization.yaml. As a user, I wonder why this is not de same for secretGenerator and configMapGenerator. Just my thoughts ;)

**Comment by sethpollack** [comment]
It is not just secretGenerator/configMapGenerator It is resources as well.

I don't think patches are getting the namespace - the resources added in `base/kustomization.yaml` are getting it and then the patches modify that.

**Comment by yanc0** [comment]
Yep, you're right. That make sense !

**Comment by StuartHarris** [comment]
We have a situation where a secret is only needed for one of the overlays (i.e. it's not in the base). That secret doesn't get the base.namespace, and I think it should.

**Comment by deedubs** [comment]
We have the use case where we are deploying multiple operators into a cluster, we have defined the operator deployments using kustomize and pull them into a clusters kustomization.yaml as bases. 

Each operator installs into its own namespace, and as such, each one needs a pull secret inserted.  Secret generator can't enumerate the namespaces so we chose to deploy a "blank" secret into each base, and then use a SecretGenerator to append the `.dockerconfigjson` attribute.  

This works except the SecretGenerator is pulling the namespaces off the secrets created in the bases and we lose the namespace we need to insert it into.  This may only be tangentially related to this particular issue but I believe its rooted in the same problem space.

Would love to get some input on how we could achieve the above, or what we can contribute to kustomize to make the above possible.
